File: Prog3/TrabalhoJAVA/src/app/Partido.java
package app;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Formatter;
import java.util.List;

/**
 * A classe Partido representa um partido na eleicao
 */
public class Partido implements Comparable<Partido> {
    static int votosLegendaTotal;

    private int numPartido;
    private int qtdCandidatosEleitos;
    private int votosNominais;
    private int votosLegenda;
    private String nomePartido;
    private String siglaPartido;
    private List<Candidato> candidatos;

    Partido(int numPartido, int votosLegenda, String nomePartido, String siglaPartido, List<Candidato> candidatos) {
        this.numPartido = numPartido;
        this.votosLegenda = votosLegenda;
        this.nomePartido = nomePartido;
        this.siglaPartido = siglaPartido;
        this.candidatos = retornaCandidatosPartido(candidatos);
        this.qtdCandidatosEleitos = retornaQtdEleitos();
        this.votosNominais = retornaQtdVotosNominais();
        votosLegendaTotal += votosLegenda;
    }

    static {
        votosLegendaTotal = 0;
    }

    /**
     * Esse metodo percorre a lista de candidatos enviada como parametro e, retorna
     * uma lista de candidatos do mesmo partido.
     * 
     * @param candidatos lista de todos candidatos
     * @return lista de candidatos do mesmo partido
     */
    private List<Candidato> retornaCandidatosPartido(List<Candidato> candidatos) {
        List<Candidato> candidatosPartido = new ArrayList<Candidato>();

        for (Candidato c : candidatos) {
            if (c.getnumPartido() == this.numPartido)
                candidatosPartido.add(c);
        }

        Collections.sort(candidatosPartido);

        return candidatosPartido;
    }

    /**
     * Esse metodo percorre a lista de candidatos local e retorna a quantidade de
     * candidatos eleitos desse partido.
     * 
     * @return quantidade de candidatos eleitos no partido
     */
    private int retornaQtdEleitos() {
        int cont = 0;
        for (Candidato c : this.candidatos) {
            if (c.getSituacao().compareTo("Eleito") == 0)
                cont++;
        }

        return cont;
    }

    /**
     * Esse metodo percorre a lista de candidatos local e retorna o total de votos
     * nominais de um partido.
     * 
     * @return quantidade de votos nominais total de um partido
     */
    private int retornaQtdVotosNominais() {
        int qtd = 0;
        for (Candidato c : this.candidatos) {
            qtd += c.getVotosNominais();
        }

        return qtd;
    }

    public int getNumPartido() {
        return numPartido;
    }

    public int getQtdCandidadosEleitos() {
        return qtdCandidatosEleitos;
    }

    public int getVotosLegenda() {
        return votosLegenda;
    }

    public String getNomePartido() {
        return nomePartido;
    }

    public String getSiglaPartido() {
        return siglaPartido;
    }

    public List<Candidato> getCandidatos() {
        return this.candidatos;
    }

    public void setNumPartido(int numPartido) {
        this.numPartido = numPartido;
    }

    public void setVotosLegenda(int votosLegenda) {
        this.votosLegenda = votosLegenda;
    }

    public void setNomePartido(String nomePartido) {
        this.nomePartido = nomePartido;
    }

    public void setSiglaPartido(String siglaPartido) {
        this.siglaPartido = siglaPartido;
    }

    @Override
    public int compareTo(Partido p) {
        int totalVotosThis = this.votosNominais + this.votosLegenda;
        int totalVotosOutro = p.votosNominais + p.votosLegenda;
        if (totalVotosOutro - totalVotosThis == 0) {
            return this.numPartido - p.numPartido;
        }
        return totalVotosOutro - totalVotosThis;
    }

    @Override
    public String toString() {
        int votosTotais = this.votosLegenda + this.votosNominais;
        Formatter fmt = new Formatter();
        fmt.format("%s - %d, %d %s (%d %s e %d de legenda), %d %s", this.siglaPartido, this.numPartido, votosTotais,
                votosTotais <= 1 ? "voto" : "votos", this.votosNominais,
                this.votosNominais <= 1 ? "nominal" : "nominais", this.votosLegenda, this.qtdCandidatosEleitos,
                this.qtdCandidatosEleitos <= 1 ? "candidato eleito" : "candidatos eleitos");
        String s = fmt.out().toString();
        fmt.close();
        return s;
    }

    public String toStringCandidatos() {
        if (this.candidatos.size() <= 0)
            return null;

        Candidato primeiro = this.candidatos.get(0);
        Candidato ultimo = this.candidatos.get(this.candidatos.size() - 1);

        Formatter fmt = new Formatter();
        fmt.format("%s - %d, %s (%d, %d %s) / %s (%d, %d %s)", this.siglaPartido, this.numPartido,
                primeiro.getNomeUrna(), primeiro.getNumero(), primeiro.getVotosNominais(),
                primeiro.getVotosNominais() <= 1 ? "voto" : "votos", ultimo.getNomeUrna(), ultimo.getNumero(),
                ultimo.getVotosNominais(), ultimo.getVotosNominais() <= 1 ? "voto" : "votos");
        String s = fmt.out().toString();
        fmt.close();

        return s;
    }

}

class ComparadorPrimeiroCandidato implements Comparator<Partido> {
    @Override
    public int compare(Partido partido1, Partido partido2) {
        int tamListaCandidatosPartido1 = partido1.getCandidatos().size();
        int tamListaCandidatosPartido2 = partido2.getCandidatos().size();

        if (tamListaCandidatosPartido1 > 0 && tamListaCandidatosPartido2 > 0) {
            Candidato primeiroP1 = partido1.getCandidatos().get(0);
            Candidato primeiroP2 = partido2.getCandidatos().get(0);

            int comp = primeiroP2.getVotosNominais() - primeiroP1.getVotosNominais();
            if (comp == 0) {
                return partido1.getNumPartido() - partido2.getNumPartido();
            }
            return comp;
        } else if (tamListaCandidatosPartido1 > 0 && !(tamListaCandidatosPartido2 > 0)) {
            return -1;
        }
        return 1;
    }
}

File: Prog3/TrabalhoC++/partido.cpp
#include "partido.h"

#include <iostream>

int Partido::votos_legenda_total = 0;

Partido::Partido() {
}

Partido::Partido(int num_partido, int votos_legenda, const string& nome_partido,
                 const string& sigla_partido, const list<Candidato*>& candidatos) {
    this->num_partido = num_partido;
    this->votos_legenda = votos_legenda;
    this->nome_partido = nome_partido;
    this->sigla_partido = sigla_partido;
    this->candidatos = retorna_candidatos_partido(candidatos);
    this->qtd_candidatos_eleitos = retorna_qtd_eleitos();
    this->votos_nominais = retorna_qtd_votos_nominais();
    votos_legenda_total += votos_legenda;
}

//lembrar de arruamr referencia no final do trabalho
list<Candidato*> Partido::retorna_candidatos_partido(const list<Candidato*>& candidatos) {
    list<Candidato*> candidatos_partido;

    for (Candidato* c : candidatos) {
        if (c->get_num_partido() == this->num_partido)
            candidatos_partido.push_back(c);
    }

    return candidatos_partido;
}

int Partido::retorna_qtd_eleitos() const {
    int cont = 0;
    for (Candidato* c : this->candidatos) {
        if (!c->get_situacao().compare("Eleito"))
            cont++;
    }

    return cont;
}

int Partido::retorna_qtd_votos_nominais() const {
    int qtd = 0;
    for (Candidato* c : this->candidatos) {
        qtd += c->get_votos_nominais();
    }

    return qtd;
}

int Partido::get_num_partido() const {
    return this->num_partido;
}

int Partido::get_qtd_candidados_eleitos() const {
    return this->qtd_candidatos_eleitos;
}

int Partido::get_votos_legenda() const {
    return this->votos_legenda;
}

string Partido::get_nome_partido() const {
    return this->nome_partido;
}

string Partido::get_sigla_partido() const {
    return this->sigla_partido;
}

list<Candidato*> Partido::get_candidatos() const {
    return this->candidatos;
}

bool Partido::comparador_primeiro_candidato(const Partido& esq, const Partido& dir) {
    int tam_lista_candidatos_partido1 = esq.get_candidatos().size();
    int tam_lista_candidatos_partido2 = dir.get_candidatos().size();

    if (tam_lista_candidatos_partido1 > 0 && tam_lista_candidatos_partido2 > 0) {
        Candidato* primeiro_p1 = esq.get_candidatos().front();
        Candidato* primeiro_p2 = dir.get_candidatos().front();

        int comp = primeiro_p2->get_votos_nominais() - primeiro_p1->get_votos_nominais();
        if (comp == 0) {
            return esq.get_num_partido() < dir.get_num_partido();
        }
        return primeiro_p2->get_votos_nominais() < primeiro_p1->get_votos_nominais();
    } else if (tam_lista_candidatos_partido1 > 0 && !(tam_lista_candidatos_partido2 > 0)) {
        return false;
    }
    return true;
}

bool Partido::operator<(const Partido& p) const {
    int total_votos_this = this->votos_nominais + this->votos_legenda;
    int total_votos_outro = p.votos_nominais + p.votos_legenda;
    if (total_votos_outro - total_votos_this == 0) {
        return this->num_partido < p.num_partido;
    }
    return total_votos_outro < total_votos_this;
}

string Partido::to_string() const {
    int votos_totais = this->votos_legenda + this->votos_nominais;
    string sub1 = this->sigla_partido + " - " + std::to_string(this->num_partido) + ",";
    string sub2 = " " + std::to_string(votos_totais) + " " + (votos_totais <= 1 ? "voto" : "votos");
    string sub3 = " (" + std::to_string(this->votos_nominais) + " " + (this->votos_nominais <= 1 ? "nominal" : "nominais");
    string sub4 = " e " + std::to_string(this->votos_legenda) + " de legenda),";
    string sub5 = " " + std::to_string(this->qtd_candidatos_eleitos) + (this->qtd_candidatos_eleitos <= 1 ? " candidato eleito" : " candidatos eleitos");
    string saida = sub1 + sub2 + sub3 + sub4 + sub5;

    return saida;
}

string Partido::to_string_candidatos() const {
    if (this->candidatos.size() <= 0)
        return "";

    Candidato* primeiro = this->candidatos.front();
    Candidato* ultimo = this->candidatos.back();

    string str1 = this->sigla_partido + " - " + std::to_string(this->num_partido) + ", " + primeiro->get_nome_urna();
    string str2 = " (" + std::to_string(primeiro->get_numero()) + ", " + std::to_string(primeiro->get_votos_nominais());
    string str3 = (primeiro->get_votos_nominais() <= 1 ? (" voto) / ") : (" votos) / ")) + ultimo->get_nome_urna();
    string str4 = " (" + std::to_string(ultimo->get_numero()) + ", " + std::to_string(ultimo->get_votos_nominais());
    string str5 = (ultimo->get_votos_nominais() <= 1 ? " voto)" : " votos)");
    string saida = str1 + str2 + str3 + str4 + str5;

    return saida;
}

Partido::~Partido() {
}